### PR TITLE
fix: correctly configure `module` and `noEmit` tsconfig and jsconfig properties

### PR DESCRIPTION
**Pull request by eltigerchino**
fixes https://github.com/sveltejs/kit/issues/11023
fixes https://github.com/sveltejs/kit/issues/9552

This PR fixes the errors found in tsconfig.json when scaffolding a typescript library project then running `pnpm check`.

> Cannot write file '/src/lib/index.js' because it would overwrite input file.

This one is fixed by providing a index.ts file instead of .js in the example lib folder. EDIT: and adding `noEmit` to tsconfig

> Option 'module' must be set to 'NodeNext' when option 'moduleResolution' is set to 'NodeNext'.

~This one is fixed by adding the `module: NodeNext` to the library project tsconfig.json and jsconfig.json files.~ EDIT: setting module: bundler

Thanks to @hyunbinseo 

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [ ] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [x] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [x] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

### CODE REVIEW

**Comment by dummdidumm** [comment]
> This one is fixed by providing a index.ts file instead of .js in the example lib folder.

This won't help though if you're creating a library using JSDoc, you won't have a `.ts` file then. This is related to #9552 and the fix is something like this: https://github.com/sveltejs/kit/issues/9552#issuecomment-1488505074

**Comment by eltigerchino** [comment]
> This won't help though if you're creating a library using JSDoc, you won't have a `.ts` file then. This is related to #9552 and the fix is something like this: [#9552 (comment)](https://github.com/sveltejs/kit/issues/9552#issuecomment-1488505074)

Based on https://vitejs.dev/guide/features.html#transpile-only and https://www.typescriptlang.org/tsconfig#noEmit it would seem that `noEmit` is the right option here as per your comment in #9552 . I don't think it will affect Vite as they use esbuild instead of tsc to transpile to Javascript. This comment https://github.com/sveltejs/kit/issues/9552#issuecomment-1757489494 also suggests that other Vite based templates are using `noEmit`.

A bit off-topic: reading Vite's docs about `tsc --noEmit` makes me wonder if we should have type checking as part of the default npm build script before `vite build`.

**Comment by eltigerchino** [comment]
I've added `noEmit` to the templates but wonder what's the difference between putting it there and putting it in the generated one in `.svelte-kit`

**Comment by dummdidumm** [comment]
The generated one means that people updating SvelteKit get the fix, while putting it in `create-svelte` means people only get it when starting a new project (or they manually update their tsconfig).
I think in this instance it makes sense to have it in the generated tsconfig. I can't imagine a case where someone would use that same tsconfig to start a completely different build where the setting actually makes a difference. If Vite ignores it in all instances that matter, then it's probably fine. If we're paranoid/unsure if there couldn't be a case for not wanting it, then it should go into `create-svelte` like you did now.

**Comment by benmccann** [comment]
Setting `noEmit` in the generated file makes sense to me. 

I'm not sure about `"moduleResolution": "NodeNext"` though. Won't that start requiring users to have a `.js` extension? In SvelteKit 2, I think we'd like to set it to `bundler` and require TypeScript 5

**Comment by eltigerchino** [comment]
> Setting `noEmit` in the generated file makes sense to me assuming we can do it just for JS projects.

I think it might be needed for TS projects too for whatever reason someone wants to have JS files in their TS project.

> I'm not sure about `"moduleResolution": "NodeNext"` though. Won't that start requiring users to have a `.js` extension? In SvelteKit 2, I think we'd like to set it to `bundler` and require TypeScript 5

`"moduleResolution": "NodeNext"` is only being set for library projects at the moment. I don't much about what `module` and `moduleResolution` do.

**Comment by dummdidumm** [comment]
Yeah it is. While we're at it, I think it makes sense to set `"moduleResolution": "bundler"` in `create-svelte` for new projects, too. In SvelteKit 2 we can make that the generated default, but to not break things we shouldn't change that in 1.x

**Comment by eltigerchino** [comment]
> While we're at it, I think it makes sense to set `"moduleResolution": "bundler"` in `create-svelte` for new projects, too. 

Is that for all projects or just the library skeleton?

**Comment by benmccann** [comment]
> "moduleResolution": "NodeNext" is only being set for library projects at the moment. 

Ah, thanks. I missed that. That sounds good to me.

> Is that for all projects or just the library skeleton?

I think libraries should be `nodenext` and applications should be `bundler`. (I usually see these written all lowercase as the TypeScript docs have them that way)

File: packages/create-svelte/shared/+skeletonlib+checkjs/jsconfig.json [hunk 1]
```diff
@@ -9,6 +9,6 @@
 		"skipLibCheck": true,
 		"sourceMap": true,
 		"strict": true,
-		"moduleResolution": "NodeNext"
+		"moduleResolution": "bundler"
```
Comment (dummdidumm) [inline packages/create-svelte/shared/+skeletonlib+checkjs/jsconfig.json:12]: sorry for the confusion - these should stay as `NodeNext`. I meant to set `bundler` for the skeleton app and showcase app, not the lib

**Comment by hyunbinseo** [comment]
> > This one is fixed by providing a index.ts file instead of .js in the example lib folder.
> 
> This won't help though if you're creating a library using JSDoc, you won't have a `.ts` file then. This is related to #9552 and the fix is something like this: [#9552 (comment)](https://github.com/sveltejs/kit/issues/9552#issuecomment-1488505074)

The reason for proposing `index.ts` is to make it match the Skeleton project template.

As of `create-svelte@5.2.0` there is this inconsistency.

| src/lib  | JSDoc    | TypeScript |
| -------- | -------- | ---------- |
| Library  | index.js | index.js * |
| Skeleton | index.js | index.ts   |


The `noEmit` related issue is mentioned in the o.g. issue.

- https://github.com/sveltejs/kit/issues/11023
- https://github.com/sveltejs/kit/issues/9552#issuecomment-1757489494

File: packages/kit/src/core/sync/write_tsconfig.js [hunk 2]
```diff
@@ -141,6 +141,7 @@ export function get_tsconfig(kit, include_base_url) {
 			lib: ['esnext', 'DOM', 'DOM.Iterable'],
 			moduleResolution: 'node',
```
Comment (dummdidumm) [inline packages/kit/src/core/sync/write_tsconfig.js:142]:
```suggestion
			moduleResolution: 'node', // TODO change to "bundler" in SvelteKit v2
```

**Review by dummdidumm — APPROVED** [review]

**Pull request merged**